import "protocol.proto";

package OstProto;
// IPv4
message Ip4 {

	enum IpAddrMode {
		e_im_fixed = 0;
		e_im_inc_host = 1;
		e_im_dec_host = 2;
		e_im_random_host = 3;
	}

	optional bool is_override_ver	= 1;
	optional bool is_override_hdrlen = 2;
	optional bool is_override_totlen = 3;
	optional bool is_override_cksum	= 4;

	optional uint32	ver_hdrlen = 5 [default = 0x45]; 
	optional uint32	tos = 6;
	optional uint32	totlen = 7;
	optional uint32	id = 8 [default = 1234];
	// TODO: rename flags to frag_flags
	optional uint32	flags = 9;
	optional uint32	frag_ofs = 10;
	optional uint32	ttl = 11 [default = 127];
	optional uint32	proto = 12;
	optional uint32	cksum = 13;

	// Source IP
	optional fixed32	src_ip = 14;
	optional IpAddrMode	src_ip_mode = 15 [default = e_im_fixed];
	optional uint32		src_ip_count = 16 [default = 16];
	optional fixed32	src_ip_mask = 17 [default = 0xFFFFFF00];
	
	// Destination IP
	optional fixed32	dst_ip = 18;
	optional IpAddrMode	dst_ip_mode = 19 [default = e_im_fixed];
	optional uint32		dst_ip_count = 20 [default = 16];
	optional fixed32	dst_ip_mask = 21 [default = 0xFFFFFF00];

	// TODO: Options
}

extend Protocol {
	optional Ip4 ip4 = 130;
}
